### ISSUE TITLE
Add Subtitle Toggle

### ISSUE THREAD

**Issue by WowbaggersLiquidLunch** [iina/iina#2856]
**What you want IINA to do:**

Add a menu item `Toggle Subtitles` to the `Subtitles` menu. The menu item will have a keyboard shortcut `S` or `⌥⌘S`. Using this menu item will show or hide currently selected subtitles.

**What IINA does currently:**

IINA currently has a `Cycle Subtitles` that switches current subtitle to the next option in the first `Subtitle` submenu. It skips the next subtitle if it's already selected in the `Second Subtitle` submenu.

**Why you think this should be added:**

Often in a video that has mostly clear speeches, except for a few seconds of ambiguous or indistinct sound, it's quite a hassle, and immersion-breaking, to go through the `Subtitles` menu to switch between the desired subtitle and `None` many times for a total of a few seconds of the video. It's a much friendlier experience to provide the user a way to quickly go back and toggle the subtitle: `←` `S`.

Another use case is for multi-lingual videos, where the user might be familiar with some of the languages spoken within, but not all of them. The user might need subtitles for parts of a video wherein the languages are unfamiliar to him, but prefers not to see subtitles in the rest of the video.

The current `Cycle Subtitle` menu option covers this feature when there is only 1 subtitle, and if it's not selected in the `Second Subtitle` submenu. However in situations typical for a Blu-ray release of a blockbuster movie, where there are more than a handful of subtitles available in the video, cycling through them all just to toggle a subtitle on/off isn't ideal.

**Examples of other projects that have something similar:**
[MPlayerX](http://mplayerx.org) has this feature.

**Comment by WowbaggersLiquidLunch** [comment]
I want to work on this. I looked through the source code, but can't quite figure out how the mapping of menu items work. 

What do I need to do in addition to adding these 2 as follows?

```swift
extension MainMenuActionHandler {
    @objc func menuToggleSubtitles(_ sender: NSMenuItem) {
        //  implementation details omitted...
    }
}
```

```swift
func updateKeyEquivalentsFrom(_ keyBindings: [KeyMapping]) {
    var settings: [(NSMenuItem, Bool, [String], Bool, ClosedRange<Double>?, String?)] = [
        //  existing elements omitted...
        (toggleSubtitles, true, ["toggle-sub"], true, nil, nil),
        //  existing elements omitted...
    ]
    //  existing implementation details omitted...
}

**Comment by lhc70000** [comment]
Thanks for your detailed feature suggestion.

This can be achieved by adding a key binding to "Cycle subtitle visibility," as mpv provided the `sub-visibility` option (actually, it's available in the default "Movist" "VLC" or "mpv" key configurations). The reason why it doesn't exist in the default IINA key configuration is that numerous users pressed this shortcut (it was the "v" key) in earlier versions and then submitted bug reports saying "subtitles disappeared".

![image](https://user-images.githubusercontent.com/8478049/78814044-9bc21d00-7a00-11ea-843a-b35a0aed782a.png)

However, I'm interested in adding it to the main menu and would like to explain more about the menu system.

There are basically 5 parts you need to work on:

1. You are right about adding the implementation in `MainMenuActionHandler`. Calling the mpv command directly should suffice:
    ```swift
    player.mpv.command(.cycle, args: ["sub-visibility"])
    ```

2. The menu item should update its title by showing something like "Show subtitles" and "Hide subtitles" according to the current state, similar to the `pause` menu item. The logic should be added to the `updateSubMenu` method.

3. The keybinding system in IINA is a bit complicated since it needs to assign shortcut keys to most menu items based on a config file. If a key binding in the config file meets the requirement of a menu item, it will be assigned to the menu item. The `settings` array in `updateKeyEquivalentsFrom()` specifies these requirements.

    The format of the `settings` array is indicated a few lines below:
    ```
    (menuItem, isIINACmd, actions, normalizeLastNum, numRange, l10nKey)
    ```

    Therefore you should insert something like
    ```swift
    (
    toggleSubtitles,  // the menu item
    false,  // it can be done with a mpv command
    ["cycle", "sub-visibility"],  // the mpv command to match
    false, nil, nil  // no need to do extra localization
    )
    ```

4. You should add the corresponding localization key (maybe `menu.show_sub` and `menu.hide_sub`) in `Localizable.strings`, at least for the `Base` and `English` languages.

5. You should also add the actual key binding to the "IINA default" key binding scheme (the other 3 schemes already contains it), under the `Assets/config` group in Xcode. The shortcut, however, should not be something that could be easily triggered by accident.

A PR would be good enough with these changes. However, ideally, IINA must show OSD messages to notify users when `sub-visibility` changed. I can implement that part later, but you can also have a try if you want :D

**Comment by WowbaggersLiquidLunch** [comment]
Thanks for this detailed list of where the changes go.

After much delay, I've finally started working on it. 

I'm going to try calling the mpv command with the `sub-visibility` argument, but I have some confusion over how it works exactly:

1. Does it work only on the first subtitle or both subtitles? I assume it works on both subtitles, but I can't find any documentation on it. Perhaps I just didn't look hard enough for it.

2. Does it show/hide the subtitle(s) with them still selected? Or, does it switch the subtitle(s) between the user-selected one(s) and `None`?

3. I found this snippet of code in `MPVOption.Subtitles` struct:

> ```swift
> /** --sub-visibility */
> static let subVisibility = "sub-visibility"
> /** --no-sub-visibility */
> static let noSubVisibility = "no-sub-visibility"
> ```

Does this mean I need to use both of them?

```swift
player.mpv.command(.cycle, args: ["sub-visibility"])
player.mpv.command(.cycle, args: ["no-sub-visibility"])
```

**Comment by WowbaggersLiquidLunch** [comment]
> The reason why it doesn't exist in the default IINA key configuration is that numerous users pressed this shortcut (it was the "v" key) in earlier versions and then submitted bug reports saying "subtitles disappeared".

Thinking back on this, I'm quite certain mpv's `sub-visibility`'s behaviour is hiding the display of subtitles, but with those hidden subtitles still selected in the menu, not changed to `None`. It makes sense that this behaviour is confusing to some users.

In this case, I want to implement 2 behaviours of showing/hiding subtitles, customisable by the user through `Preferences`: 

- The opt-in behaviour is that of mpv's.

- The default behaviour is the opposite, similar to that of MPlayerX's:

    - When the user chooses to **hide** subtitles, they are switched to `None`, and IINA remembers what they were previously.

    - When the user chooses to **show** subtitles, they are restored to their previous selections.

    - If both subtitles are `None`, they are automatically recognised as hidden.

    - If subtitles are hidden, and both subtitles' previous selections were `None`, when the user chooses to **show** subtitles, the first subtitle will select the first subtitle track, while the second subtitle will remain as `None`.

Regardless of whether subtitles are hidden or shown when the user changes this subtitle toggle's behaviour in `Preferences`, the transition will be seamless:

- If subtitles are hidden: 

    - Going from default to mpv behaviour: Subtitles' selections are restored but they are not displayed.

    - Going from mpv behaviour to default: Subtitles are selected to `None`, but they are displayed.

- If subtitles are shown, then there are no additional changes needed.

If the maintainers consider this a suitable implementation of the subtitle toggle, I need guidance on navigating the code structure for adding a new `Preferences` option. 

As an aside, is there a particular reason why the `Preference` type is defined as such:

```swift
struct Preference {
    struct Key: RawRepresentable, Hashable {
        typealias RawValue = String
        var rawValue: RawValue
        var hashValue: Int {
            return rawValue.hashValue
        }
    }
    static let somePreference = Key("somePreference")
    //  more static properties of Preference.Key type...
}
```

instead of this: 

```swift
enum Preference: String, Hashable {
    case somePreference
    //  more cases...
}
```

**Comment by lhc70000** [comment]
Sorry for my late reply! I'm not sure whether you are still interested, but here are some quick answers:

> Does this mean I need to use both of them?

No, we use mpv _properties_ rather than _options_ when executing commands, but we are using the option struct because most options are available as properties. `--no-abc` is an alias of `--abc=no`.

> As an aside, is there a particular reason why the Preference type is defined as such:

No, it's unclear why I chose to do so back in 2017 when developing the initial version of this App while learning Swift. A lot of parts need to be refactored, but currently I have no time.

Besides, your idea looks reasonable and I'm totally agree with it. This action should be an IINA command (please check `IINACommand.swift` and `toggle-flip` as an example) so it can be saved in the key binding conf file.

**Comment by unique-username313** [comment]
@lhc70000 I don't know whether @WowbaggersLiquidLunch is still working on this or not. I have written a toggle for subtitle visibility that functions exactly like the default mpv behavior. It doesn't set subtitles to "none" and will not change upon selecting no subtitles. It is accessible from the Subtitles menu and from the Quicksettings pane.
To avoid the confusion that you talked about, several measures were taken:

- The toggle is named "Hide Subtitles". It's a switch in the subtitle quicksettings menu and there's also a checkmark in the subs menu item that appears whenever activated. That makes the current state of subtitle visibility easily seen.
<img width="306" alt="iina5" src="https://user-images.githubusercontent.com/32626712/85914428-210dde80-b846-11ea-8e75-3e95cfaa5f9a.png">


- the switch in the Quicksettings pane is positioned at the top just above the list of subtitles in order to be seen whenever a subtitle is to be changed.
<img width="1072" alt="iina2" src="https://user-images.githubusercontent.com/32626712/85914419-0a678780-b846-11ea-97f8-4d5b64b9c2ed.png">

- Pressing a keyboard shortcut to toggle sub visibility, clicking on "Hide Subtitles" in the subs menu, or toggling the switch in the Quicksettings pane will result in an OSD message showing the state of the toggle, e.g. "Hide Subtitles: On" or "Hide Subtitles: Off"
<img width="1072" alt="iina3" src="https://user-images.githubusercontent.com/32626712/85914393-d8562580-b845-11ea-9dc3-9813d35def23.png">

I don't think that messing with what the toggle does by default like settings subs to none or choosing the first sub is not a good idea. The first subtitle is usually English and a user might want another language to be selected by default. Well we can make the toggle check for the preferred sub language in the preferences but that adds another level of checks and variables that is unnecessary . Also maintaining two different behaviors for the same functionality is also cumbersome. All of this increases the complexity of a feature that is supposed have a very simple function: Hide Subtitles, or don't Hide Subtitles. The default mpv behavior is perfect and now with an OSD message and the feedback from the toggle and menu item, user confusion is less likely.

My implementation could be improved by making it a IINA command with a default key binding, but I haven't done that since I use my own mpv bindings and my use case doesn't demand it.

In case @WowbaggersLiquidLunch is not wokring on this anymore and the maintainers feel that such an implementation is suitable, I'll make a PR. Please let me know.

**Comment by WowbaggersLiquidLunch** [comment]
I'm still working on it in principle, although I did pause the coding bit while waiting for a response previously. I'm also kind of waiting for #2877, because I don't like working with CocoaPods. In any case, I think I will get back on implementing it after this weekend.

@unique-username313 I think what you implemented is great. It gives the user multiple places for hiding/unhiding subtitles, which is essential for a good user experience. Also as you said, your implementation is consistent with mpv's behaviour. 

That being said, I don't think mpv's behaviour itself is the best for user experience. And I think my proposed approach is better than mpv's. 

---

Let's examine the following scenario where the user interacts with subtitles, one using mpv's default behaviour, the other using my proposed:

### 1. mpv's default:

The user starts a video that has multiple subtitle tracks and one audio track. The audio track contains both English and French speeches. The user doesn't need subtitles for most of the video, because he understands both languages. However, he needs the subtitles from time to time for a few seconds each, because the speech isn't always clear.

A few minutes into the video, the user encounters the first few seconds of unclear English, so he unwinds by 10 seconds and selects the English subtitle. Right afterwards, the user **hides** the the subtitle, because he doesn't think he will need it again. 

10 minutes later, there is another few seconds of unclear English, so he seeks back by 10 seconds, **unhides** the subtitle, watch the bit with unclear English with subtitle, then **hides** it again.

5 minutes later, a similar thing happens, but this time it's in French. So the user unwinds the video, goes to the Subtitles menu, and selects French. But this time, the subtitle doesn't appear, because it's **hidden**. The user is puzzled at first, then remembers and **unhides** the subtitle. Then he watches, **hides** the subtitle, and continues watching.

### 2. my proposed solution:

A few minutes into the video, the user encounters the first few seconds of unclear English, so he unwinds by 10 seconds and selects the English subtitle. Right afterwards, the user **turns off** the the subtitle, because he doesn't think he will need it again. 

10 minutes later, there is another few seconds of unclear English, so he seeks back by 10 seconds, **turns on** the subtitle, watch the bit with unclear English with subtitle, then **turns off** it again.

5 minutes later, a similar thing happens, but this time it's in French. So the user unwinds the video, goes to the Subtitles menu, and selects French. And there it is, the French subtitle on screen. So he watches, then **turns off** the subtitle, and continues watching.

---

The key difference here is that mpv changes the subtitle's visibility, while my proposed approach changes the subtitles' selection. 

What I see in mpv's behaviour is that although it gets the job down on the surface, it requires the user's knowledge that it's only the visibility that changes. However, the user is more often than not incognisant of it. As illustrated in the hypothetic scenario, this creates a fair bit of friction. And it's this kind of friction (no matter how small) that separate a good user experience from a bad one.

Another argument for my proposed approach is that, since there already exists a "none" selection that perfectly represents the "off" state for subtitles, then why not use it directly? It only complicates the app's internal logic by adding 2 extra state variables that track the 2 subtitles, and the entire behaviour is transparent to the user with no surprises.

I completely agree with your argument that

> maintaining two different behaviors for the same functionality is cumbersome.

And this is exactly a problem that mpv's default behaviour has. It adds subtitle visibility on top of the none/some selection. This is parallel to wrapping optional over an optional type in Swift. It just creates confusion. 

My solution, on the other hand, provides clarity, because it uses the "none" selection directly, and adds no ambiguity.

---

There is another, previously unstated, reason for my pushing for an alternative approach from mpv's default. I envision adding Siri support to IINA after this, with a feature that allows the user to unwind the video and momentarily turn on the subtitle just by saying "Hey Siri, what did he say?", exactly like to how it works on Apple TV. This feature will be much better and easier built on my proposed approach than on mpv's default. 

---

Let's debate.

**Comment by lhc70000** [comment]
@unique-username313, thanks for your work and it looks great. Are you using the newly added `Switch` view?

I'm OK with either solution. I still believe that it's better to switch between tracks rather than touching `sub-visibility`, but I think @unique-username313's design is clear enough and useful for users who need such a function.

**Comment by unique-username313** [comment]
@lhc70000 Thank you for the reply, and yes, I'm using the same custom switch view that was added for hwdec and deinterlace.

@WowbaggersLiquidLunch thanks a lot for your comprehensive reply. Although I still don't agree with all of your points, what I was proposing was only the poor man's alternative for the original feature request, and it was only an option in the event that you dropped this project.

> This is parallel to wrapping optional over an optional type in Swift. It just creates confusion.

I have to disagree with you here. It's not wrapping an optional over an optional type. These are two different concepts. The first is --sid: the currently selected subtitle, the other is a flag: subtitle visibility, regardless of --sid. This is exactly the same way muting audio works:

- Let's say you press the mute button. Muting audio doesn't set the audio track to none, it just toggles the sound which is _the way you perceive_ audio.
When you press the mute toggle again, it switches sound back on, but the property --aid has never changed, it was never set to none.

- In parallel to that, when you toggle sub-visibility, you toggle the text rendering on and off, which is _the way you perceive_ text, regardless of which subtitle was selected before that.

We fundamentally have different features in mind, I have a toggle to _mute_ subtitles, you're making a toggle to _disable_ subtitles. And this is why debating which is best and which is confusing **by comparing them** is not really fruitful. I agree that _muting_ subtitles can be a confusing concept and this exactly why I wrote the switch, OSD message, and menu button (for something I always used with just the keybind "v") after I read this feature request and saw this:

> The reason why it doesn't exist in the default IINA key configuration is that numerous users pressed this shortcut (it was the "v" key) in earlier versions and then submitted bug reports saying "subtitles disappeared"

I'm not debating you which behavior is better than the other because I believe that we're talking about two different concepts. Like I said above, I'm not pushing for a certain way or the other, my suggestion was only an alternative in case you stopped working on this. It's true that it wasn't exactly what you suggested, but it was a simple enough toggle to write and suggest to @lhc70000 here. But all of this is moot now since I know you'll be back on this.

**Comment by WowbaggersLiquidLunch** [comment]
> It's not wrapping an optional over an optional type. These are two different concepts.

You're right. It's a poor analogy I made.

> But all of this is moot now since I know you'll be back on this.

@lhc70000 What do you think if we implement both @unique-username313's and mine, and add a section in the preferences for the user to choose which behaviour they want? I believe @unique-username313's implementation adds value. It improves upon the previously disabled "v" keyboard shortcut, but providing GUI controls and informs the user on changes.

**Comment by WowbaggersLiquidLunch** [comment]
I just realised that having both behaviours is actually what I proposed in April.

> In this case, I want to implement 2 behaviours of showing/hiding subtitles, customisable by the user through Preferences:
> 
> - The opt-in behaviour is that of mpv's.
> 
> - The default behaviour is the opposite, similar to that of MPlayerX's:

**Comment by unique-username313** [comment]
> Overall, I guess what I'm trying to say is that, although hiding/unhiding subtitles and muting/unmuting audios are parallel in their presentation and underlying mechanisms, they are very different in use cases from the user's perspective.

You’re absolutely right. I do believe that the feature you want to add is fundamentally different from muting subtitles. What I did with my code was just integrating an already present mpv feature in the OSD and UX, what you’re thinking about is a downright new feature.

What I was saying is that sub-visibility is a toggle to mute subtitles, and what you want has nothing to do with it.

I thought about your feature and how it could be implemented in IINA, and there is actually no need at all to use mpv’s sub-visibility property (actually an option that is also a property at runtime). It’s a matter of switching  the values of sid and secondary-sid between none and lastused (even if lastused was already none).

You also need to watch sid and secondary-sid for changes (in MPVController.swift), and update your toggle function according to those changes.

> @lhc70000 What do you think if we implement both @unique-username313's and mine, and add a section in the preferences for the user to choose which behaviour they want?

I don’t think that’s necessary. What I did isn’t even _implementing anything_ now that think about it. Switching sub-visibility was always there to be used with a key binding. All I did was prettify the UX around that “already present” feature. Users who want to switch sub-visibility on and off can alway add the appropriate binding. Your feature is actually neat and @lhc70000 already said that he prefers it to be implemented.

On the other hand, in case my changes are welcome, I think that letting the user choose between the two is a good idea, but the toggles and menu items should be renamed accordingly. Either “Hide/Show Subtitles” or “Disable/Enable Subtitles”. OSD messages should also change according to the preference: either “Subtitles Hidden/Shown” or “Subtitles Enabled/Disabled”

**Comment by unique-username313** [comment]
@WowbaggersLiquidLunch okay so I got bored yesterday and tried to write it your way as a small exercise. Now I’m not a developer and I don’t even work in IT so the code might be **very** ugly, but it works. Here is the commit: https://github.com/unique-username313/iina/commit/446e66ecb8bac73c6c1fdb5f60a77956a864d148

Since you didn’t mention a Quicksettings toggle in your description of the feature above, I didn’t put it there.

You’re welcome to check the commit out and use it as you see fit.

**Comment by WowbaggersLiquidLunch** [comment]
> Now I’m not a developer and I don’t even work in IT so the code might be **very** ugly

You're too humble.

I think you code looks good, although I haven't tested it yet. Would you mind opening a pull request to [my fork][WowbaggersLiquidLunch/iina]? This way, instead of me stealing your code, your contribution will be recorded by Git?

[WowbaggersLiquidLunch/iina]: https://github.com/WowbaggersLiquidLunch/iina

**Comment by lhc70000** [comment]
If any of you are going to open a PR, could you only submit the mpv/menu/key binding logic and leave the UI part (if any) to me? There are a lot of UI changes waiting to be merged in v1.1, and I'm worried about conflicts, as conflicts in xibs can be hard to resolve.

@WowbaggersLiquidLunch You can cherry-pick @unique-username313's commit to preserve the original author.

**Comment by unique-username313** [comment]
^ my last try to
implement this that I worked on has no UI changes whatsoever. @WowbaggersLiquidLunch took over the code. You got me intrigued by those UI changes now :)

**Comment by WowbaggersLiquidLunch** [comment]
> leave the UI part

Does it include OSD, menubar and/or contextual menus? I suppose so long as we don't implement custom UI styles, everything should just adopt whatever UI changes there are in v1.1?

> There are a lot of UI changes waiting to be merged in v1.1

Just out of curiosity, are the changes mostly Big Sur related?

> You can cherry-pick

Thanks. I totally forgot about this.

**Comment by unique-username313** [comment]
@WowbaggersLiquidLunch I just remembered that there's a menu item in MainMenu.xib. I guess that should be reverted in the code I submitted to you.

**Comment by WowbaggersLiquidLunch** [comment]
Sorry for the big delay. I finally found time to actually resume working on it today, and I plan to open a draft pull request in a few hours. 

@unique-username313, I see you left a `FIXME` comment on [line 662 in `PlayerCore.swift`](https://github.com/WowbaggersLiquidLunch/iina/blob/083f2cd6a3a4b070c0e786db67192dd17e2eec8c/iina/PlayerCore.swift#L662):

> sendOSD doesn't work here because as soon as a sub track changes, the property listeners in MPVController will fire their own OSD message effectively making the ones here unseen

@lhc70000 A few questions regarding this: Is there currently a way for OSD messages to be aware of other OSD messages? If there isn't one at the moment, can I put all OSD messages on a sequential asynchronous dispatch queue, so they can queue up and wait for their turns? Should this be implemented in a separate pull request? Or, should I skip all OSD changes altogether until the new UI changes are merged in v1.1?

**Comment by WowbaggersLiquidLunch** [comment]
> I just remembered that there's a menu item in MainMenu.xib. I guess that should be reverted in the code I submitted to you.

I think I found the one. If it's the missing `disableSubtitle` menu item, then I think I fixed it.

**Comment by unique-username313** [comment]
> After looking through the commit history, I don't think I've found it.

https://github.com/WowbaggersLiquidLunch/iina/blob/083f2cd6a3a4b070c0e786db67192dd17e2eec8c/iina/Base.lproj/MainMenu.xib#L514-L516

https://github.com/WowbaggersLiquidLunch/iina/blob/083f2cd6a3a4b070c0e786db67192dd17e2eec8c/iina/Base.lproj/MainMenu.xib#L695

**Comment by unique-username313** [comment]
>  I see you left a FIXME

Sorry about that, that FIXME was for me before you took over and I forgot about it. I remember at the time that I thought of a way to work around the issue mentioned by implementing a logic in handlePropertyChange in those two cases

``` Swift
    case MPVOption.TrackSelection.sid:
      player.info.sid = Int(getInt(MPVOption.TrackSelection.sid))
      player.postNotification(.iinaSIDChanged)
      player.sendOSD(.track(player.info.currentTrack(.sub) ?? .noneSubTrack))

    case MPVOption.Subtitles.secondarySid:
      player.info.secondSid = Int(getInt(MPVOption.Subtitles.secondarySid))
      player.postNotification(.iinaSIDChanged)
```

and choose which OSDMessage should be sent according to the toggle state. But then I thought that it was too hacky. So I left it and forgot to remove all the OSD related stuff before pushing a PR.

**Comment by WowbaggersLiquidLunch** [comment]
> > After looking through the commit history, I don't think I've found it.
> 
> Those are the lines I was talking about
> 
> https://github.com/WowbaggersLiquidLunch/iina/blob/083f2cd6a3a4b070c0e786db67192dd17e2eec8c/iina/Base.lproj/MainMenu.xib#L514-L516
> 
> https://github.com/WowbaggersLiquidLunch/iina/blob/083f2cd6a3a4b070c0e786db67192dd17e2eec8c/iina/Base.lproj/MainMenu.xib#L695

Ok. Then I did find and fix it. Thanks!

Btw, I am to blame for the missing menu item. I left it hanging there when I first started working on this in April.

**Comment by WowbaggersLiquidLunch** [comment]
> > I see you left a FIXME
> 
> Sorry about that, that FIXME was for me before you took over and I forgot about it. 

Actually I'm quit glad that you left it there, because it exposes an OSD problem that we never discussed: When a user (re-)enables the subtitles, should the OSD message be "Subtitles enabled", or "Subtitle: # 1 English", or "Subtitles enabled: # 1 English"? Or, should there be 2 separate OSD messages "Subtitles enabled" and "Subtitle: # 1 English", either appearing simultaneously or successively?

> I remember at the time that I thought of a way to work around the issue mentioned by implementing a logic in handlePropertyChange in those two cases
> 
> ```swift
>     case MPVOption.TrackSelection.sid:
>       player.info.sid = Int(getInt(MPVOption.TrackSelection.sid))
>       player.postNotification(.iinaSIDChanged)
>       player.sendOSD(.track(player.info.currentTrack(.sub) ?? .noneSubTrack))
> 
>     case MPVOption.Subtitles.secondarySid:
>       player.info.secondSid = Int(getInt(MPVOption.Subtitles.secondarySid))
>       player.postNotification(.iinaSIDChanged)
> ```
> 
> and choose which OSDMessage should be sent according to the toggle state. But then I thought that it was too hacky. So I left it and forgot to remove all the OSD related stuff before pushing a PR.

I think this way should work if the player always displays only 1 OSD message. We would just need to check `savedSid` and `savedSecondSid` (which I renamed to [`recoveredFirstSubtitleTrackIndex`](https://github.com/iina/iina/blob/a928feef48ca8371cd57da393448bfb96e19bb8e/iina/PlaybackInfo.swift#L162) and [`recoveredSecondSubtitleTrackIndex`](recoveredSecondSubtitleTrackIndex)) and figure out whether the subtitles' state just changed.

Alternatively, we can move the subtitle-change's OSD logic to [`sid`](https://github.com/iina/iina/blob/a928feef48ca8371cd57da393448bfb96e19bb8e/iina/PlaybackInfo.swift#L138)'s and [`secondSid`](https://github.com/iina/iina/blob/a928feef48ca8371cd57da393448bfb96e19bb8e/iina/PlaybackInfo.swift#L149)'s observers, so we might be able to avoid checking twice whether the subtitles' state just changed.

@lhc70000 What's your opinion?

**Comment by unique-username313** [comment]
I guess there is no need to display any OSD message at all in regards to the toggle status since the already-present OSD messages will imply the underlying toggle status.

- If OSD says `Subtitle: # 1 English`, it implies that subs are enabled.
- if OSD says `Subtitle: # 1 None` (or whatever the message is, I don't have access to IINA right now), it implies that subs are disabled by the toggle.

This differs from the visibility toggle. That one needs a message since tracks aren't being switched around.

**Comment by WowbaggersLiquidLunch** [comment]
Here is another previously unconsidered (at least unconsidered by me) case for the track-switching toggle: If the user uses both the primary and secondary subtitles, is there a way for the player to display changes to both subtitles simultaneously in a vertical stack?

**Comment by unique-username313** [comment]
Well stock IINA doesn't display any OSD message upon switching `--secondary-sid`. Only `--sid`.

I'm guessing here but I think that this is intentional on the part of the maintainers because implementing this would mean simultaneous OSD messages, which means stacks of messages one on top of another, or consecutive messages.

- the former would be screen clutter, taking valuable screen estate.
- the latter would contradict the purpose of OSD messages, which are supposed to be brief, non obtrusive, and descriptive of an event that *just happened*, not "a few moments earlier". Consecutive messages are distracting and can be confusing because, like I said, would appear late compared to the event they're describing.

That's my opinion about it anyway, the IINA team might have a different idea though.
